### PR TITLE
Chapter 13: User Microposts

### PR DESCRIPTION
**Pull request by puteriffah**

### CODE REVIEW

**Comment by puteriffah** [comment]
ready

File: app/assets/stylesheets/microposts.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Microposts controller here.
```
Comment (manhnt-0960) [inline app/assets/stylesheets/microposts.scss]: remove empty file

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -135,3 +135,57 @@ input.btn.btn-primary {
     border-bottom: 1px solid $gray-lighter;
   }
 }
+
+/* Microposts */
+aside {
+
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]:
```suggestion
```

File: app/assets/stylesheets/custom.scss [hunk 3]
```diff
@@ -135,3 +135,57 @@ input.btn.btn-primary {
...
+  }
+}
+
+.microposts {
+  list-style: none;
+  padding: 0;
+
+  li {
+  padding: 10px 0;
+  border-top: 1px solid #e8e8e8;
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: indenting 2 spaces

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -17,4 +17,12 @@ def extract_locale
   def default_url_options
     {locale: I18n.locale}
   end
+
+  def logged_in_user
+    return if logged_in?
+
+      store_location
+      flash[:danger] = t "login.title"
+      redirect_to login_path
```
Comment (manhnt-0960) [inline app/controllers/application_controller.rb]: pls check indent again,

File: app/assets/stylesheets/custom.scss [hunk 5]
```diff
@@ -135,3 +135,57 @@ input.btn.btn-primary {
...
+  }
+
+  .user {
+  margin-top: 5em;
+  padding-top: 0;
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: and this

File: app/assets/stylesheets/microposts.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/microposts.scss]: remove file that never used

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (saobv-2223) [inline app/controllers/application_controller.rb]:
```suggestion
    store_location
    flash[:danger] = t "login.title"
    redirect_to login_path
```

File: app/controllers/microposts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  def micropost_params
+    params.require(:micropost).permit :content, :image
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    return if @micropost
+
+      flash[:warning] = t "micropost.notfound"
+      redirect_to root_path
```
Comment (manhnt-0960) [inline app/controllers/microposts_controller.rb]:
```suggestion
    flash[:warning] = t "micropost.notfound"
    redirect_to root_path
```

File: app/controllers/microposts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
+class MicropostsController < ApplicationController
+  before_action :logged_in_user, only: %i(create destroy)
+  before_action :correct_user, only: :destroy
+
+  def create
+    @micropost = current_user.microposts.build micropost_params
+    @micropost.image.attach params[:micropost][:image]
+    if @micropost.save
+
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb]:
```suggestion
```

File: app/controllers/microposts_controller.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      flash[:success] = t "micropost.create.success"
+      redirect_to root_path
+    else
+      @feed_items = current_user.feed.page params[:page]
+
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb]:
```suggestion
```

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -12,7 +12,9 @@ def new
     @user = User.new
   end
 
-  def show; end
+  def show
+    @microposts = @user.microposts.paginate page: params[:page]
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]: add number of per page

File: app/controllers/microposts_controller.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  def destroy
+    if @micropost.destroy
+      flash[:success] = t "micropost.delete.success"
+    else
+      flash[:warning] = t "micropost.delete.fail"
+    end
+    redirect_back fallback_location: root_path
+  end
+
+    private
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb]:
```suggestion
  private
```

File: app/controllers/microposts_controller.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    params.require(:micropost).permit :content, :image
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    return if @micropost
+
+      flash[:warning] = t "micropost.notfound"
+      redirect_to root_path
+    end
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb]:
```suggestion
    flash[:warning] = t "micropost.notfound"
    redirect_to root_path
```

File: app/helpers/microposts_helper.rb [hunk 13]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/microposts_helper.rb]: remove file that never used

File: app/models/micropost.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  scope :recent_posts, ->{order created_at: :desc}
+  has_one_attached :image
+  validates :user_id, presence: true
+  validates :content, presence: true, length: {maximum: Settings.validations.content.max_length}
+  validates :image, content_type: { in: Settings.file.image_type, message: I18n.t("errors.messages.invalid_pic") }, size: { less_than: Settings.file.picture_size.megabytes, message: I18n.t("errors.messages.less_than") }
```
Comment (saobv-2223) [inline app/models/micropost.rb]:
```suggestion
  belongs_to :user
  
  scope :recent_posts, ->{order created_at: :desc}
  
  has_one_attached :image
  
  validates :user_id, presence: true
  validates :content, presence: true, length: {maximum: Settings.validations.content.max_length}
  validates :image, content_type: { in: Settings.file.image_type, message: I18n.t("errors.messages.invalid_pic") }, size: { less_than: Settings.file.picture_size.megabytes, message: I18n.t("errors.messages.less_than") }
```

File: app/views/static_pages/home.html.erb [hunk 15]
```diff
@@ -1,7 +1,17 @@
...
+      <section class="micropost_form">
+        <%= render "shared/micropost_form" %>
+      </section>
+    </aside>
+    <div class="col-md-8">
+      <h3><%= t "micropost.feed" %></h3>
+      <%= render "shared/feed" %>
+    </div>
+  </div>
+<% else %>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: if else case does not have anything, delete it.
```suggestion
```

File: app/models/micropost.rb [hunk 16]
```diff
@@ -0,0 +1,12 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  scope :recent_posts, ->{order created_at: :desc}
```
Comment (manhnt-0960) [inline app/models/micropost.rb]: move line 3 to line 8

File: app/views/static_pages/home.html.erb (hunk 15 above)
Comment (manhnt-0960) [inline app/views/static_pages/home.html.erb]: delete this line

**Comment by puteriffah** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**